File: webui/src/runtime/com/sun/webui/jsf/component/Accordion.java
/*
 * The contents of this file are subject to the terms
 * of the Common Development and Distribution License
 * (the License).  You may not use this file except in
 * compliance with the License.
 * 
 * You can obtain a copy of the license at
 * https://woodstock.dev.java.net/public/CDDLv1.0.html.
 * See the License for the specific language governing
 * permissions and limitations under the License.
 * 
 * When distributing Covered Code, include this CDDL
 * Header Notice in each file and include the License file
 * at https://woodstock.dev.java.net/public/CDDLv1.0.html.
 * If applicable, add the following below the CDDL Header,
 * with the fields enclosed by brackets [] replaced by
 * you own identifying information:
 * "Portions Copyrighted [year] [name of copyright owner]"
 * 
 * Copyright 2007 Sun Microsystems, Inc. All rights reserved.
 */

package com.sun.webui.jsf.component;

import com.sun.faces.annotation.Property;
import com.sun.faces.annotation.Component;
import com.sun.webui.jsf.theme.ThemeImages;
import com.sun.webui.jsf.util.ComponentUtilities;
import com.sun.webui.jsf.util.ThemeUtilities;
import com.sun.webui.theme.Theme; 

import javax.el.ValueExpression;
import javax.el.ValueExpression;
import javax.faces.component.UIComponent;
import javax.faces.context.FacesContext;

/**
 * An accordion container. It extends the TabContainer and adds some 
 * functionality that is specific to the Accordion. An accordion can be
 * thought of as a vertical tab set. It can contain one of more accordion tabs
 * each of which can contain virtually any HTML markup. In general accordions
 * are used for navigational purposes - each tab contains links which when 
 * clicked takes the user to different sections (tasks) of the application.
 * The Accordian allows one or more tabs to be open at any given time. When the
 * accordion is configured to allow multiple tabs to be open at any given time 
 * it supports "expandAll" and "collapseAll" icons. The expandAll icon when clicked
 * will expand all tabs, and colapseAll, will collapse all open tabs. 
 * An accordion can be refreshed. 
 * Refreshing the accordion will cause it to render itself and all its children
 * again. The Accordion can also be refreshed to go through all the steps of 
 * the JSF lifecycle as opposed to just the render response phase.
 * It is adviseable to use an alternate navigational component if the number of
 * tabs in the accordion are exceeding ten.
 */
@Component(type="com.sun.webui.jsf.Accordion", 
    family="com.sun.webui.jsf.Accordion", displayName="Accordion", 
    tagName="accordion", tagRendererType="com.sun.webui.jsf.widget.Accordion")
public class Accordion extends TabContainer {
    
    /**
     * Create a new Accordion.
     */
    public Accordion() {
        super();
        setRendererType("com.sun.webui.jsf.widget.Accordion");
    }
    
    /**
     * <p>Return the family for this component.</p>
     */
    public String getFamily() {
        return "com.sun.webui.jsf.Accordion";
    }

    /**
     * <p>Return the renderer type associated with this component.</p>
     */
    public String getRendererType() {
        if (ComponentUtilities.isAjaxRequest(getFacesContext(), this)) {
            return "com.sun.webui.jsf.ajax.Accordion";
        }
        return super.getRendererType();
    }
      
    /**
     * Returns true if multiple tabs can be selected at the same time.
     * By default this setting is false and only one accordion tab can be
     * selected at any given time. Note that when only a single tab is
     * selected the accordion will not supply expand/collapse icons even if 
     * the application developer provides facets for these.
     */
    @Property(name="multipleSelect", displayName="Multiple tab selected")
    private boolean multipleSelect = false;
    private boolean multipleSelect_set = false;
    
    /**
     * Returns true if multiple tabs can be selected, false otherwise.
     * This value is false by default.
     */
    public boolean isMultipleSelect() {
        if (this.multipleSelect_set) {
            return this.multipleSelect;
        }
        ValueExpression _vb = getValueExpression("multipleSelect");
        if (_vb != null) {
            Object _result = _vb.getValue(getFacesContext().getELContext());
            if (_result == null) {
                return false;
            } else {
                return ((Boolean) _result).booleanValue();
            }
        }
        return false;
    }
    
    /**
     * Set to true if multiple tabs can be selected. 
     */
    public void setMultipleSelect(boolean multipleSelect) {
        this.multipleSelect = multipleSelect;
        this.multipleSelect_set = true;
    }

    /**
     * Set to true if the accordion container should display
     * expandAll and collapseAll control features. Note that this setting applies
     * only in the case where multipleSelect has been set to true. In the case
     * of single select accordions only one tab can be open at any given time.
     */
    @Property(name="toggleControls", displayName="Set expand and colapse controls")
    private boolean toggleControls = false;
    private boolean toggleControls_set = false;
    
    /**
     * Returns true if the Accordion container contains expand/collapse
     * controls. 
     */
    public boolean isToggleControls() {
        if (this.toggleControls_set) {
            return this.toggleControls;
        }
        ValueExpression _vb = getValueExpression("toggleControls");
        if (_vb != null) {
            Object _result = _vb.getValue(getFacesContext().getELContext());
            if (_result == null) {
                return false;
            } else {
                return ((Boolean) _result).booleanValue();
            }
        }
        return false;
    }
    
    /**
     * Set to true if Accordion container should contain expand/collapse
     * controls. 
     */
    public void setToggleControls(boolean toggleControls) {
        this.toggleControls = toggleControls;
        this.toggleControls_set = true;
    }
    
    
    /**
     * Should be set to true if the accordion container is to display
     * a refresh icon. A refresh icon would generally be provided on
     * the accordion header. When clicked it would refresh the accordion
     * and its tabs asynchronously.
     */
    @Property(name="refreshIcon", displayName="Display Refresh Icon")
    private boolean refreshIcon = false;
    private boolean refreshIcon_set = false;
    
    /**
     * Returns true if the Accordion container contains a refresh icon.
     * 
     */
    public boolean isRefreshIcon() {
        if (this.refreshIcon_set) {
            return this.refreshIcon;
        }
        ValueExpression _vb = getValueExpression("refreshIcon");
        if (_vb != null) {
            Object _result = _vb.getValue(getFacesContext().getELContext());
            if (_result == null) {
                return false;
            } else {
                return ((Boolean) _result).booleanValue();
            }
        }
        return false;
    }
    
    /**
     * Set to true if Accordion container should contain a refresh icons.
     * 
     */
    public void setRefreshIcon(boolean refreshIcon) {
        this.refreshIcon = refreshIcon;
        this.refreshIcon_set = true;
    }
    
    /**
     * Restore the state of this component.
     */
    @Override
    public void restoreState(FacesContext _context,Object _state) {
        Object _values[] = (Object[]) _state;
        super.restoreState(_context, _values[0]);
        this.multipleSelect = ((Boolean) _values[1]).booleanValue();
        this.multipleSelect_set = ((Boolean) _values[2]).booleanValue();
        this.toggleControls = ((Boolean) _values[3]).booleanValue();
        this.toggleControls_set = ((Boolean) _values[4]).booleanValue();
        this.refreshIcon = ((Boolean) _values[5]).booleanValue();
        this.refreshIcon_set = ((Boolean) _values[6]).booleanValue();
    }
    
    /**
     * Save the state of this component.
     */
    @Override
    public Object saveState(FacesContext _context) {
        Object _values[] = new Object[7];
        _values[0] = super.saveState(_context);
        _values[1] = this.multipleSelect ? Boolean.TRUE : Boolean.FALSE;
        _values[2] = this.multipleSelect_set ? Boolean.TRUE : Boolean.FALSE;
        _values[3] = this.toggleControls ? Boolean.TRUE : Boolean.FALSE;
        _values[4] = this.toggleControls_set ? Boolean.TRUE : Boolean.FALSE;
        _values[5] = this.refreshIcon ? Boolean.TRUE : Boolean.FALSE;
        _values[6] = this.refreshIcon_set ? Boolean.TRUE : Boolean.FALSE;
        return _values;
    }
   
}

File: themes/javascript/widget/accordion.js
//
// The contents of this file are subject to the terms
// of the Common Development and Distribution License
// (the License).  You may not use this file except in
// compliance with the License.
// 
// You can obtain a copy of the license at
// https://woodstock.dev.java.net/public/CDDLv1.0.html.
// See the License for the specific language governing
// permissions and limitations under the License.
// 
// When distributing Covered Code, include this CDDL
// Header Notice in each file and include the License file
// at https://woodstock.dev.java.net/public/CDDLv1.0.html.
// If applicable, add the following below the CDDL Header,
// with the fields enclosed by brackets [] replaced by
// you own identifying information:
// "Portions Copyrighted [year] [name of copyright owner]"
// 
// Copyright 2007 Sun Microsystems, Inc. All rights reserved.
//

dojo.provide("webui.@THEME@.widget.accordion");

dojo.require("dojo.widget.*");
dojo.require("webui.@THEME@.widget.*");

/**
 * This function is used to generate a template based widget.
 *
 * Note: This is considered a private API, do not use.
 */
webui.@THEME@.widget.accordion = function() {
    // Register widget.
    dojo.widget.HtmlWidget.call(this);
}

/**
 * Helper function to add accordion header controls
 */
webui.@THEME@.widget.accordion.addControls = function(props) {       
    // Add expand and collapse icons only if multiple select is set to
    // true and the icons have been supplied.
    if (props.toggleControls && props.multipleSelect) {
        // Set expand all image properties.
        if (props.expandAllImage) {
            // Set properties.
            props.expandAllImage.id = this.expandAllImage.id; // Required for updateFragment().

            // Update/add fragment.
            this.widget.updateFragment(this.expandAllImgContainer, props.expandAllImage);
        }

        // Set collapse all image properties.
        if (props.collapseAllImage) {
            // Set properties.
            props.collapseAllImage.id = this.collapseAllImage.id; // Required for updateFragment().

            // Update/add fragment.
            this.widget.updateFragment(this.collapseAllImgContainer, props.collapseAllImage);
        }
    }

    // Set refresh image properties.
    if (props.refreshImage) {
        // Set properties.
        props.refreshImage.id = this.refreshImage.id; // Required for updateFragment().

        // Update/add fragment.
        this.widget.updateFragment(this.refreshImgContainer, props.refreshImage);
    }
    return true;
}

/**
 * Close all open accordions and leave the others as is.
 */
webui.@THEME@.widget.accordion.collapseAllTabs = function() {
    for (var i=0; i<this.children.length; i++) {
        var child = this.children[i];
        if (child.widgetType == "accordionTab") {
            if (child.selected) {
                child.setSelected(false);
            }
        }
    }
    return true;
}

/**
 * This closure is used to process widget events.
 */
webui.@THEME@.widget.accordion.event = {
    /**
     * This closure is used to process refresh events.
     */
    refresh: {
        /**
         * Event topics for custom AJAX implementations to listen for.
         */
        beginTopic: "webui_@THEME@_widget_accordion_event_refresh_begin",
        endTopic: "webui_@THEME@_widget_accordion_event_refresh_end"
    },

    /**
     * This closure is used to process state change events.
     */
    state: {
        /**
         * Event topics for custom AJAX implementations to listen for.
         */
        beginTopic: "webui_@THEME@_widget_accordion_event_state_begin",
        endTopic: "webui_@THEME@_widget_accordion_event_state_end"
    }
}

/**
 * Open all closed tabs and leave the others as is.
 */
webui.@THEME@.widget.accordion.expandAllTabs = function() {
    for (var i=0; i<this.children.length; i++) {
        var child = this.children[i];
        if (child.widgetType == "accordionTab") {
            if (!child.selected) {
                child.setSelected(true);
            }
        }
    }
    return true;
}

/**
 * This function is used to fill in template properties.
 *
 * Note: This is called after the buildRendering() function. Anything to be set 
 * only once should be added here; otherwise, use the _setProps() function.
 *
 * @param props Key-Value pairs of properties.
 * @param frag HTML fragment.
 */
webui.@THEME@.widget.accordion.fillInTemplate = function(props, frag) {
    webui.@THEME@.widget.accordion.superclass.fillInTemplate.call(this, props, frag);

    with (this.domNode.style) {
        if (position != "absolute") {
            position = "relative";
        }
        overflow = "hidden";
    }

    // Set ids.
    if (this.id) {
        this.domNode.id = this.id;
        this.headerContainer.id = this.id + "_accHeader";
        this.expandAllContainer.id = this.id + "_expandAllNode";
        this.expandAllImgContainer.id = this.expandAllContainer.id + "_expandAllImage";
        this.collapseAllContainer.id = this.id + "_collapseAllNode";
        this.collapseAllImgContainer.id = this.collapseAllImgContainer.id + "_collapseAllImage";
        this.dividerNodeContainer.id = this.id + "_dividerNode";
        this.refreshNodeContainer.id = this.id + "_refreshNode";
    }

    // Set class names.
    this.collapseAllContainer.className = webui.@THEME@.widget.props.accordionTab.accordionHdrCloseApp;
    this.dividerNodeContainer.className = webui.@THEME@.widget.props.accordionTab.accordionHdrDivider;
    this.expandAllContainer.className = webui.@THEME@.widget.props.accordionTab.accordionHdrOpenAll;
    this.refreshNodeContainer.className = webui.@THEME@.widget.props.accordionTab.accordionHdrRefresh;

    // Set events.
    var id = this.id;
    dojo.event.connect(this.collapseAllContainer, "onclick", this, "collapseAllTabs");
    dojo.event.connect(this.expandAllContainer, "onclick", this, "expandAllTabs");
    dojo.event.connect(this.refreshNodeContainer, "onclick", function() {
        var widget = dojo.widget.byId(id);
        widget.refresh(id);
    });

    return true;
}

/**
 * This function is used to get widget properties. Please see the 
 * _setProps() function for a list of supported properties.
 */
webui.@THEME@.widget.accordion.getProps = function() {
    var props = webui.@THEME@.widget.accordion.superclass.getProps.call(this);

    // Set properties.
    if (this.loadOnSelect) { props.loadOnSelect = this.loadOnSelect; }
    if (this.toggleControls) { props.toggleControls = this.toggleControls; }
    if (this.multipleSelect) { props.multipleSelect = this.multipleSelect; }
    if (this.expandAllImage != null) { props.expandAllImage = this.expandAllImage; }
    if (this.collapseAllImage != null) { props.collapseAllImage = this.collapseAllImage; }
    if (this.refreshImage != null) { props.refreshImage = this.refreshImage; }
    if (this.style != null) { props.style = this.style; }
    if (this.styleClass != null) { props.styleClass = this.styleClass; }
    if (this.tabs != null) { props.tabs = this.tabs; }
    if (this.id) { props.id = this.id; }
    if (this.type) { props.type = this.type; }

    return props;
}

/**
 * This function selects the child tab when the user clicks
 * on its label. The actual behavior of the accordion depends on
 * whether multipleSelect is enabled or not.
 */
webui.@THEME@.widget.accordion.selectChild = function(widget) {
    if (this.multipleSelect) {
        widget.setSelected(true);
    } else {
        for (var i=0; i<this.children.length; i++) {
            var child = this.children[i];
            if (child.widgetType == "accordionTab") {
                child.setSelected(child.id == widget.id);
            }
        }
    }
}

/**
 * This function is used to set widget properties. Please see the 
 * _setProps() function for a list of supported properties.
 *
 * Note: This function updates the widget object for later updates. Further, the
 * widget shall be updated only for the given key-value pairs.
 *
 * Note: If the notify param is true, the widget's state change event shall be
 * published. This is typically used to keep client-side state in sync with the
 * server.
 *
 * @param props Key-Value pairs of properties.
 * @param notify Publish an event for custom AJAX implementations to listen for.
 */
webui.@THEME@.widget.accordion.setProps = function(props, notify) {
    if (props == null) {
        return false;
    }

    // Replace tabs -- do not extend.
    if (props.tabs) {
        this.tabs = null;
    }

    // Extend widget object for later updates.
    return webui.@THEME@.widget.accordion.superclass.setProps.call(this, props, notify);
}

/**
 * This function is used to set widget properties with the following 
 * Object literals.
 *
 * <ul>
 *  <li>id</li>
 *  <li>loadOnSelect</li>
 *  <li>toggleControls</li>
 *  <li>multipleSelect</li>
 *  <li>expandAllImage</li>
 *  <li>collapseAllImage</li>
 *  <li>refreshImage</li>
 *  <li>style</li>
 *  <li>styleClass</li>
 *  <li>tabs</li>
 *  <li>type</li>
 * </ul>
 *
 * Note: This is considered a private API, do not use. This function should only
 * be invoked through postInitialize() and setProps(). Further, the widget shall
 * be updated only for the given key-value pairs.
 *
 * @param props Key-Value pairs of properties.
 */
webui.@THEME@.widget.accordion._setProps = function(props) {
    if (props == null) {
        return false;
    }

    // add control icons - refresh, expandall, collapseall.
    this.addControls(props); 

    // If we are coming here for
    // the first time there will be no children. The other case is when 
    // the accordion is being rerendered because of a refresh in which 
    // we want to use the latest set of children. addFragment is supposed
    // to do that.
    if (props.tabs) {
        // To Do: Before adding new children, previously created nodes should be
        // removed via the removeChildNodes(domNode) function -- see label. 
        //
        // Adding children to a specific container (other than this.domNode) 
        // would help isolate what needs to be removed. The following code works
        // because widgets of the same id are destroyed before creating new objects.
        for (var i=0; i < props.tabs.length; i++) {
            this.widget.addFragment(this.domNode, props.tabs[i], "last");
        }
    }

    // Set more properties..
    this.setCommonProps(this.domNode, props);
    this.setEventProps(this.domNode, props);

    // Set remaining properties.
    return webui.@THEME@.widget.accordion.superclass._setProps.call(this, props);
}

// Inherit base widget properties.
dojo.inherits(webui.@THEME@.widget.accordion, webui.@THEME@.widget.widgetBase);

// Override base widget by assigning properties to class prototype.
dojo.lang.extend(webui.@THEME@.widget.accordion, {
    // Set private functions.
    addControls: webui.@THEME@.widget.accordion.addControls,
    collapseAllTabs: webui.@THEME@.widget.accordion.collapseAllTabs,
    expandAllTabs: webui.@THEME@.widget.accordion.expandAllTabs,
    fillInTemplate: webui.@THEME@.widget.accordion.fillInTemplate,
    getProps: webui.@THEME@.widget.accordion.getProps,
    selectChild: webui.@THEME@.widget.accordion.selectChild,
    setProps: webui.@THEME@.widget.accordion.setProps,
    _setProps: webui.@THEME@.widget.accordion._setProps,

    // Set defaults.
    duration: 250,
    event: webui.@THEME@.widget.accordion.event,
    isContainer: true,
    loadOnSelect: false,
    multipleSelect: false,
    widgetType: "accordion"
});
